pkg/dockerregistry/server/manifestservice.go

```diff
@@ -0,0 +1,245 @@
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	quotautil "github.com/openshift/origin/pkg/quota/util"
+)
+
+var _ distribution.ManifestService = &manifestService{}
+
+type manifestService struct {
+	ctx       context.Context
+	repo      *repository
+	manifests distribution.ManifestService
+
+	// acceptschema2 allows to refuse the manifest schema version 2
+	acceptschema2 bool
+}
+
+// Exists returns true if the manifest specified by dgst exists.
+func (m *manifestService) Exists(ctx context.Context, dgst digest.Digest) (bool, error) {
+	context.GetLogger(ctx).Debugf("(*manifestService).Exists")
+
+	image, err := m.repo.getImage(dgst)
+	if err != nil {
+		return false, err
+	}
+	return image != nil, nil
+}
+
+// Get retrieves the manifest with digest `dgst`.
+func (m *manifestService) Get(ctx context.Context, dgst digest.Digest, options ...distribution.ManifestServiceOption) (distribution.Manifest, error) {
+	context.GetLogger(ctx).Debugf("(*manifestService).Get")
+
+	if _, err := m.repo.getImageStreamImage(dgst); err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving ImageStreamImage %s/%s@%s: %v", m.repo.namespace, m.repo.name, dgst.String(), err)
+		return nil, err
+	}
+
+	image, err := m.repo.getImage(dgst)
+	if err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving image %s: %v", dgst.String(), err)
+		return nil, err
+	}
+
+	ref := imageapi.DockerImageReference{
+		Namespace: m.repo.namespace,
+		Name:      m.repo.name,
+		Registry:  m.repo.registryAddr,
+	}
+	if managed := image.Annotations[imageapi.ManagedByOpenShiftAnnotation]; managed == "true" {
+		// Reference without a registry part refers to repository containing locally managed images.
+		// Such an entry is retrieved, checked and set by blobDescriptorService operating only on local blobs.
+		ref.Registry = ""
+	} else {
+		// Repository with a registry points to remote repository. This is used by pullthrough middleware.
+		ref = ref.DockerClientDefaults().AsRepository()
+	}
+
+	manifest, err := m.manifests.Get(WithRepository(ctx, m.repo), dgst, options...)
+	switch err.(type) {
+	case distribution.ErrManifestUnknownRevision:
+		break
+	case nil:
+		m.repo.rememberLayersOfManifest(manifest, ref.Exact())
+		m.updateManifest(ctx, image)
+		return manifest, nil
+	default:
+		context.GetLogger(ctx).Errorf("unable to get manifest from storage: %v", err)
+		return nil, err
+	}
+
+	if len(image.DockerImageManifest) == 0 {
+		// We don't have the manifest in the storage and we don't have the manifest
+		// inside the image so there is no point to continue.
+		return nil, distribution.ErrManifestUnknownRevision{
+			Name:     m.repo.Named().Name(),
+			Revision: dgst,
+		}
+	}
+
+	manifest, err = m.repo.manifestFromImageWithCachedLayers(image, ref.Exact())
+
+	if _, err := m.manifests.Put(WithRepository(ctx, m.repo), manifest, options...); err != nil {
```

Comment:
This should be in a goroutine and use a map to check uniqueness (exactly like my blob mirroring PR).  If we need to factor to common code that's fine.